sdk-rs/src/lib.rs

```diff
@@ -0,0 +1,790 @@
+                &[],
+                &[],
+            );
+
+            let ix = Instruction {
+                program_id: constants::PROGRAM_ID,
+                accounts,
+                data: InstructionData::data(&drift_program::instruction::ModifyOrder {
+                    order_id: Some(order_id),
+                    modify_order_params: ModifyOrderParams {
+                        direction: Some(params.direction),
+                        base_asset_amount: Some(params.base_asset_amount),
+                        price: Some(params.price),
+                        max_ts: params.max_ts,
+                        oracle_price_offset: params.oracle_price_offset,
+                        auction_duration: params.auction_duration,
+                        auction_end_price: params.auction_end_price,
+                        auction_start_price: params.auction_start_price,
+                        trigger_price: params.trigger_price,
+                        ..Default::default()
+                    },
+                }),
+            };
+            self.ixs.push(ix);
+        }
+
+        self
+    }
+
+    /// Build the transaction ready for signing and sending
+    pub fn build(self) -> Transaction {
+        Transaction::new_with_payer(self.ixs.as_ref(), Some(&self.wallet.authority()))
+    }
+}
+
+/// Builds a set of required accounts from a user's open positions and additional given accounts
+///
+/// `base_accounts` base anchor accounts
+///
+/// `user` Drift user account data
+///
+/// `markets_readable` IDs of markets to include as readable
+///
+/// `markets_writable` ` IDs of markets to include as writable (takes priority over readable)
+///
+/// # Panics
+///  if the user has positions in an unknown market (i.e unsupported by the SDK)
+fn build_accounts(
+    context: Context,
+    base_accounts: impl ToAccountMetas,
+    user: &User,
+    markets_readable: &[MarketId],
+    markets_writable: &[MarketId],
+) -> Vec<AccountMeta> {
+    // the order of accounts returned must be instruction, oracles, spot, perps see (https://github.com/drift-labs/protocol-v2/blob/master/programs/drift/src/instructions/optional_accounts.rs#L28)
+    let mut seen = [0_u64; 2]; // [spot, perp]
+    let mut accounts = Vec::<AccountType>::default();
+
+    // add accounts to the ordered list
+    let mut include_market = |market_index: u16, market_type: MarketType, writable: bool| {
+        let index_bit = 1_u64 << market_index as u8;
+        // always safe since market type is 0 or 1
+        let seen_by_type = unsafe { seen.get_unchecked_mut(market_type as usize % 2) };
+        if *seen_by_type & index_bit > 0 {
+            return;
+        }
+        *seen_by_type |= index_bit;
+
+        let (account, oracle) = match market_type {
+            MarketType::Spot => {
+                let SpotMarketConfig {
+                    account, oracle, ..
+                } = constants::spot_market_by_index(context, market_index);
```

Comment:
this will make the sdk fall over if the user interacts w sdk for a new market. probably fine for now, but we made it so drift client can derive the all the market accounts at subscription so if bot running the sdk restarts, it handles it gracefully